christfellowshipchurch/web-app
# [Events] Responsive design polish and bug fixes

# Changes

### Basecamp todos addressed
- [**🔥 [Chat] Update copy on offline state**](https://3.basecamp.com/3926363/buckets/17566548/todos/3049860513)
- [**🔥 [Bug] Unable to load an Event Chat**](https://3.basecamp.com/3926363/buckets/17566548/todos/3061631315)
- [**🔥 [Events] Better handle responsive mobile web view**](https://3.basecamp.com/3926363/buckets/17566548/todos/3058078474)
- [**🔥 [Events] Move Calls to Action out of the Panel**](https://3.basecamp.com/3926363/buckets/17566548/todos/3058078478)
- [**🏕 [Events] Animate live indicator for extra "live" feel**](https://3.basecamp.com/3926363/buckets/17566548/todos/3050194849)
- [**🏕 [Chat] useEffect for initializing chat fires back-to-back**](https://3.basecamp.com/3926363/buckets/17566548/todos/3057764922) _(attempted to fix)_
### Additional Changes
- Lots of refinement/polish to padding etc for better mobile/tablet experience
- Add `<GridContainer>` , `<Row>`, `<Col>` components and uses them in the new Events components
  - These are simple wrappers for Bootstrap classes, intended to improve code readability
- Add `breakpoint` utility and media query breakpoints from Bootstrap to `styled-components` theme
- In a few places, adds lodash `get` to reduce chance of unexpected Rock/API data from crashing page

# Tradeoffs/Caveats
Responsive layout for the `<EventPanel>`, the side view with **Schedule** and **Chat** tabs, is still clunky on device/orientation combinations like landscape phone and portrait tablet.
In general, this PR focuses on at least _having those elements visible and usable_, and I expect to have to make another smaller pass to further refine most likely after launch. IMO, current state is not release blocking.

I attempted to fix an intermittent bug that occurs when logging in while on a live event. I'm not 100% confident I "truly" fixed it, as it seems slightly quirky and like a potential async race condition when Stream client disconnects and reinitializes.
My attempted fix is to `await` the `StreamChatClient.disconnect()` call in the `useEffect`'s cleanup function, hoping it won't proceed to next effect invocation until it's done cleaning up 🤷‍♂️.

# Test Instructions
- 🔸 Verify the "chat offline" state copy by viewing any event that isn't currently live
- 🔸 Verify the empty chat state (live, but no messages) by hard-coding `'dev-empty-channel'` in place of `channelId` on `src/content-single/EventContentItem/components/EventChat.js::132`
- 🔸 Verify that you don't get an error when logging in on a live event, after loading the page in a logged out state
- 🔸 Verify general design/layout looks decent across 

# Screenshots

|🎥 <a href="https://www.loom.com/share/9c1bac5b3ecf4f26a279cb57b693feba">Loom Video Overview</a>|
|---|
|<a href="https://www.loom.com/share/9c1bac5b3ecf4f26a279cb57b693feba"><img src="https://user-images.githubusercontent.com/1812955/94718832-4b2cdc00-0320-11eb-8bf1-10cbdbc6a25b.jpg" width="420" /></a>|

|Size|Screenshot|
|---|---|
|iPhone X|<img src="https://user-images.githubusercontent.com/1812955/94719691-7b28af00-0321-11eb-8fe2-e23f87840e23.jpg" width="200" />|
|Small-ish Tablet|<img src="https://user-images.githubusercontent.com/1812955/94718960-7e6f6b00-0320-11eb-90d6-668efbd73386.jpg" width="240" />|
|Compressed Desktop/Almost Tablet|<img src="https://user-images.githubusercontent.com/1812955/94718966-80392e80-0320-11eb-8a67-ddbab017c96a.jpg" width="280" />|
|Desktop|<img src="https://user-images.githubusercontent.com/1812955/94718969-80d1c500-0320-11eb-91d7-19afecce1954.jpg" width="300" />|
|X-Large Desktop|<img src="https://user-images.githubusercontent.com/1812955/94718971-816a5b80-0320-11eb-993e-73d50f0b3e97.jpg" width="300" />|

#### Live Now indicator animation
<img src="https://user-images.githubusercontent.com/1812955/94718990-86c7a600-0320-11eb-8201-97698f51030e.gif" width="500" />

#### Chat List States
|Offline/Not Live|Live, but Empty|
|---|---|
|<img src="https://user-images.githubusercontent.com/1812955/94722606-88479d00-0325-11eb-9eeb-7cc08dc9a260.png" width="300" />|<img src="https://user-images.githubusercontent.com/1812955/94722478-5b938580-0325-11eb-877e-53fb58c4c1d7.jpg" width="300" />|


# Tests
